### PR TITLE
Issue 51: Add RSA key generation as an optional feature

### PR DESCRIPTION
**Pull request by ericeberry**
Wait for Issue 49 to be merged before reviewing

### CODE REVIEW

**Review by riwoh — DISMISSED** [review]

**Review by mhabrat — CHANGES_REQUESTED** [review]

File: reference/src/taimpl/src/internal/rsa.c [hunk 1]
```diff
@@ -575,3 +575,87 @@ sa_status rsa_sign_pss(
...
+            ERROR("EVP_PKEY_CTX_new failed");
+            break;
+        }
+
+        if (EVP_PKEY_keygen_init(evp_pkey_ctx) != 1) {
+            ERROR("EVP_PKEY_keygen_init failed");
+            break;
+        }
+
+        if (EVP_PKEY_CTX_set_rsa_keygen_bits(evp_pkey_ctx, (int) parameters->modulus_length * 8) != 1) {
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/rsa.c:617]: This is the standard way of using OpenSSL. 95% of functions return 1 as success. All of the calls to OpenSSL in all of the functions follow this pattern.

File: reference/src/taimpl/src/internal/rsa.c [hunk 2]
```diff
@@ -575,3 +575,87 @@ sa_status rsa_sign_pss(
...
+    size_t key_length = 0;
+    do {
+
+        evp_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+        if (evp_pkey_ctx == NULL) {
+            ERROR("EVP_PKEY_CTX_new failed");
+            break;
+        }
+
+        if (EVP_PKEY_keygen_init(evp_pkey_ctx) != 1) {
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/rsa.c:612]: This is the standard way of using OpenSSL. 95% of functions return 1 as success. All of the calls to OpenSSL in all of the functions follow this pattern.

File: reference/src/taimpl/src/internal/rsa.c [hunk 3]
```diff
@@ -575,3 +575,87 @@ sa_status rsa_sign_pss(
...
+            ERROR("EVP_PKEY_keygen_init failed");
+            break;
+        }
+
+        if (EVP_PKEY_CTX_set_rsa_keygen_bits(evp_pkey_ctx, (int) parameters->modulus_length * 8) != 1) {
+            ERROR("EVP_PKEY_CTX_set_ec_paramgen_curve_nid failed");
+            break;
+        }
+
+        if (EVP_PKEY_keygen(evp_pkey_ctx, &evp_pkey) != 1) {
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/rsa.c:622]: This is the standard way of using OpenSSL. 95% of functions return 1 as success. All of the calls to OpenSSL in all of the functions follow this pattern.

File: reference/src/taimpl/include/internal/rsa.h [hunk 4]
```diff
@@ -166,6 +166,19 @@ sa_status rsa_sign_pss(
         size_t in_length,
         bool precomputed_digest);
 
+/**
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/include/internal/rsa.h:169]: Done

File: reference/src/taimpl/src/internal/rsa.c [hunk 5]
```diff
@@ -575,3 +575,87 @@ sa_status rsa_sign_pss(
 
     return status;
 }
+
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/rsa.c:578]: Done

File: reference/src/taimpl/src/internal/rsa.c [hunk 6]
```diff
@@ -575,3 +575,87 @@ sa_status rsa_sign_pss(
...
+    sa_status status = SA_STATUS_INTERNAL_ERROR;
+    uint8_t* key = NULL;
+    EVP_PKEY_CTX* evp_pkey_ctx = NULL;
+    EVP_PKEY* evp_pkey = NULL;
+    size_t key_length = 0;
+    do {
+
+        evp_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
+        if (evp_pkey_ctx == NULL) {
+            ERROR("EVP_PKEY_CTX_new failed");
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/rsa.c:608]: Done

File: reference/src/taimpl/src/internal/rsa.c [hunk 7]
```diff
@@ -575,3 +575,87 @@ sa_status rsa_sign_pss(
...
+            break;
+        }
+
+        if (EVP_PKEY_keygen_init(evp_pkey_ctx) != 1) {
+            ERROR("EVP_PKEY_keygen_init failed");
+            break;
+        }
+
+        if (EVP_PKEY_CTX_set_rsa_keygen_bits(evp_pkey_ctx, (int) parameters->modulus_length * 8) != 1) {
+            ERROR("EVP_PKEY_CTX_set_ec_paramgen_curve_nid failed");
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/internal/rsa.c:618]: Done

File: reference/src/taimpl/src/ta_sa_key_generate.c [hunk 8]
```diff
@@ -146,8 +147,52 @@ static sa_status ta_sa_key_generate_rsa(
...
-    ERROR("RSA key generation is not supported");
-    return SA_STATUS_OPERATION_NOT_SUPPORTED;
+    if (key == NULL) {
+        ERROR("NULL key");
+        return SA_STATUS_NULL_PARAMETER;
+    }
+    *key = INVALID_HANDLE;
+
+    if (rights == NULL) {
+        ERROR("NULL parameters");
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_generate.c:157]: Done

File: reference/src/taimpl/src/ta_sa_key_generate.c [hunk 9]
```diff
@@ -146,8 +147,52 @@ static sa_status ta_sa_key_generate_rsa(
...
+        ERROR("NULL caller_uuid");
+        return SA_STATUS_NULL_PARAMETER;
+    }
+
+    sa_status status;
+    stored_key_t* stored_key = NULL;
+    do {
+        status = rsa_generate_key(&stored_key, rights, parameters);
+        if (status != SA_STATUS_OK) {
+            ERROR("ec_generate_key failed");
```
[the remark this answers is not in the document]
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_generate.c:181]: done

File: reference/src/taimpl/src/ta_sa_key_generate.c [hunk 10]
```diff
@@ -154,7 +154,7 @@ static sa_status ta_sa_key_generate_rsa(
     *key = INVALID_HANDLE;
 
     if (rights == NULL) {
-        ERROR("NULL parameters");
```
Comment (mhabrat) [inline reference/src/taimpl/src/ta_sa_key_generate.c]: Looks like this file needs the copyright year extension.

**Review by mhabrat — CHANGES_REQUESTED** [review]

**Review by mhabrat — APPROVED** [review]

File: reference/src/taimpl/src/ta_sa_key_generate.c (hunk 10 above)
Comment (ericeberry) [reply reference/src/taimpl/src/ta_sa_key_generate.c]: Done

**Pull request merged**